adrs/3122196229/README.md

```diff
@@ -0,0 +1,55 @@
+# NodeJS files and directories names conversion
+
+* **State:** Draft
+* **Created:** 2020-11-10
+* **Tags:** nodejs
+
+## Context
+
+Across the ecosystem, we can find all sort of files and directories naming
+conversions in NodeJS projects.
+
+That has caused misalignment, shifting the focus away from more important
+matters, also have created conventions that introduce problems such:
+
+* Disagreement around English language, making it even more difficult for the
+  non-English speaker to figure out the right decision.
+* Case sensitive vs. case insensitive file systems, causing issues in different
+  environments such as your local computer vs. CI/CD computers.
+* Format mismatching around separating English words.
+
+The `camelCase`, and/or `PascalCase` for naming convention, although this seems
+a good solution there are some caveats.
```

Comment:
We should detail the strongest arguments for `camelCase` and `PascalCase`?  In these ADRs, I hope to see some "steel manning" -- where we address the best version of the opposing arguments.

I'll take a stab: I think the strongest argument for case-sensitive filenames is that case-sensitive filenames can do a better job at describing the file contentst.  By convention, there is a meaningful difference between `camelCase`, `PascalCase`,  and `CONSTANT_CASE`.

- `Calculator.ts` is a Class
- `Calculator.tsx` is a JSX component
- `calculator.ts` is a function
- `useCaclulator.ts` is a react hook
- `CALCULATOR.ts` is some sort of constant

Because of the casing, you can make pretty good guesses about what to expect in each of these files.

Reply:
> Calculator.ts is a Class

Does that mean I can't have anything else? 

Most of these practices come from languages like C# and Java, where, as @tleef explained, it makes sense because your class is the code's container. The language itself constraints to such practices, or the ecosystem already decided the answer to my question: yes, it only contains the class's code.

In JavaScript/Node/React ecosystem, you probably have more than just the class since JavaScript files are containers, and they dont require classes to execute the code.

> Calculator.tsx is a JSX component

Does that mean I can't have anything else than a component?

The same problem as before, but around what people do, if you go for this route, we shouldn't add things like creating Context (In React ecosystem); therefore, you will now require to create a context file.

Also, the file is `.tsx` (which will be part of another ADR to use either `.tsx` or `.jsx`), so you didn't lose that context.

> calculator.ts is a function
> useCaclulator.ts is a react hook

The difference between `useCalculator` and `calculator` is the `use-` part, so it doesn't matter since the context's meaning is around the react hook. The usage of `use-,` therefore, the name has meaning, no the casing itself.

To conclude, 

The problem is that files and folders describe a correlation between content in at least two-dimensional space (Folder + File). Practices where you couple the files and folders to the content create more problems than those that solve.

- Required refactoring
- Increase entropy since you have more files and folders.
- No clear boundaries of when to create more files and folders. (Do I suppose to create one-function-one-file?)

Reply:
Worth saying, in another ADR, I will propose some conversion for the React ecosystem about `use-x.ts` file, and `context` file, and other things.

That way the cognitive load is lower, but what makes it lower is more the conversion than the casing itself.

Reply:
> Increase entropy since you have more files and folders.
This is a good point.

The aforementioned naming convention would likely encourage developers to break things up into separate files in order to keep the naming convention "true" instead of breaking things up when it makes sense.

Reply:
> > `Calculator.tsx` is a JSX component
> 
> 
> Does that mean I can't have anything else than a component?

What else besides the component would you want in that file ?


I could imagine that file having a bunch of private functions/utilities,  type definitions, and constants,  and whatever else it needs to render.

Reply:
Right, the boundaries are not clear, so most of the time those files are not just the component, and for a good reason.

The more files and folders, the higher the entropy, which entropy is not necessarily bad done right.

For example, seal a legacy code (legacy doesn't mean trash the code because it is bad, legacy means old), organize things based on responsibilities, and such since you know the contexts, and what will change at what point.

But I wouldn't recommend doing such a thing from the get-go.